### PR TITLE
Change Discover navigation & flow to be dynamic from the user's selection

### PR DESCRIPTION
**Pull request by ryanclark**
When changing between self hosted and AWS databases, we need to change the flow and navigation as they have steps that differ between them.

To do this, I've added the ability for a resource to specify its views as a function instead of a static array, so the navigation and flow can change depending on whatever the user has selected.

To do this, `useDiscover` has been changed to be a context, so `useDiscover` can be used in multiple places and share the same data. 

It's now become generic as it can hold a generic state for each resource, which is passed into the views function.

As a result, the resource definition is now generic -

```typescript
interface Database {
  location: DatabaseLocation;
}

export const DatabaseResource: Resource<Database> = {
  views(database) { // database is type Database
    if (database.location === DatabaseLocation.AWS) {
      return [ ... ];
    }

    return [ ... ];
  }
}
```

This generic state can be modified by using the discover context

```typescript
export function SelectDatabaseType() {
  const { resourceState, setResourceState } = useDiscover<Database>();

  // setResourceState can now be called with something that 
  // implements `Database`, and the navigation and flow will change 
}
```

I've also added a database context in the wrapper, so we can persist state between the different pages of the flow. We _could_ use the discover context for this, as it can now store a generic state for the resource that's been selected, but I think individual, specific typed contexts are better for each resource. The generic resource state on the discover context should be used for changing the navigation/flow.

This is what it looks like without a database type selected

image

After selecting an AWS database resource, the navigation changes

image

And the same for self hosted

image

### CODE REVIEW

File: packages/teleport/src/Discover/Database/index.tsx [hunk 1]
```diff
@@ -16,48 +16,86 @@
...
-  ],
+  views(database) {
+    let configureResourceViews;
+    if (database) {
+      switch (database.location) {
+        case DatabaseLocation.AWS:
+          configureResourceViews = [
+            {
+              title: 'Deploy Database Agent',
+              component: InstallActiveDirectory,
```
Comment (ryanclark) [inline packages/teleport/src/Discover/Database/index.tsx:48]: FYI All of these `InstallActiveDirectory` will need to be changed to the actual pages

**Review by kimlisa — APPROVED** [review]
looks good other than that we should remove the database provider for now. 

and i just wanted to double check, looks like another round of refactor is coming our way (not near, but after) and it will look something like this right:

```ts
// currently we are passing agent props to each view
function TestConnection(props: AgentStepProps) {}

// now we use context
function TestConnection() {
  const props = useDiscover()
}
```

is that right?

File: packages/teleport/src/Discover/Database/DatabaseContext.tsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+export function DatabaseProvider(props: DatabaseProviderProps) {
+  const [database, setDatabase] = useState<Database>(null);
+
+  return (
+    <databaseContext.Provider value={{ database, setDatabase }}>
+      {props.children}
+    </databaseContext.Provider>
+  );
+}
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply packages/teleport/src/Discover/Database/DatabaseContext.tsx:24]: I completely agree, I'll remove this context provider

File: packages/teleport/src/Discover/Database/resources.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    location: DatabaseLocation.AWS,
+    engine: DatabaseEngine.Redis,
+    name: 'AWS MemoryDB for Redis',
+  },
+
+  {
+    location: DatabaseLocation.SelfHosted,
+    engine: DatabaseEngine.Redis,
+    name: 'Self Hosted Redis',
+    popular: true,
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply packages/teleport/src/Discover/Database/resources.ts:56]: Sure thing

**Pull request merged**